src/runtime_src/core/tools/xbutil2/OO_P2P.cpp

```diff
@@ -17,12 +17,244 @@
 // Local - Include Files
 #include "OO_P2P.h"
 #include "tools/common/XBUtilities.h"
+#include "core/common/error.h"
+#include "core/common/unistd.h"
+#include "core/common/memalign.h"
+#include "core/common/device.h"
+#include "core/common/query_requests.h"
+#include "experimental/xrt-next.h"
+#include "experimental/xrt_bo.h"
+#include "xrt.h"
+#include "xclbin.h"
 namespace XBU = XBUtilities;
 
 // 3rd Party Library - Include Files
+#include <boost/program_options.hpp>
+#include <boost/algorithm/string/join.hpp>
+namespace po = boost::program_options;
 
 // System - Include Files
 #include <iostream>
+#include <cctype>
+
+#ifdef _WIN32
+# pragma warning(disable : 4245)
+#endif
+
+namespace {
+
+enum class action_type { enable, disable, validate };
+
+action_type
+string2action(std::string str)
+{
+  std::transform(str.begin(), str.end(), str.begin(),
+                 [](auto c) { return static_cast<char>(std::tolower(c)); });
+
+  if (str == "enable")
+    return action_type::enable;
+  else if (str == "disable")
+    return action_type::disable;
+  else if (str == "validate")
+    return action_type::validate;
+  else
+    throw xrt_core::generic_error(EINVAL, "Invalid p2p action '" + str + "'");
+}
+
+static auto
+p2p_config(xrt_core::device* device)
+{
+  try {
+    auto cfg = xrt_core::device_query<xrt_core::query::p2p_config>(device);
+    return cfg;
+  }
+  catch (const std::exception&) {
+    throw xrt_core::system_error(ENOTSUP, "p2p is not supported");
+  }
+}
+
+static void
+p2p_enabled_or_error(xrt_core::device* device)
+{
+  auto cfg = p2p_config(device);
+
+  long long bar = -1;
+  long long rbar = -1;
+  long long remap = -1;
+  long long exp_bar = -1;
+
+  for (auto& str : cfg) {
+    auto pos = str.find(":");
+    auto key = str.substr(0, pos);
+    ++pos; // past ":"
+    if (key == "bar")
+      bar = std::stoll(str.substr(pos));
+    else if (key == "exp_bar")
+      exp_bar = std::stoll(str.substr(pos));
+    else if (key == "rbar")
+      rbar = std::stoll(str.substr(pos));
+    else if (key == "remap")
+      remap = std::stoll(str.substr(pos));
+  }
+
+  if (bar == -1)
+    throw xrt_core::system_error(ENOTSUP,"p2p is not supported");
+  if (rbar != -1 && rbar > bar)
+    throw xrt_core::system_error(EIO,"Please WARM reboot to enable p2p");
+  if (remap > 0 && remap != bar)
+    throw xrt_core::system_error(EIO,"p2p remapper is not set correctly");
+  if (bar == exp_bar)
+    return;
+
+  throw xrt_core::system_error(ENOTSUP,"p2p is not supported");
+}
+
+namespace p2ptest {
```

Comment:
I added p2p test in SubCmdValidate

Reply:
Oh, I didn't know about these., thanks for pointing that out.  Unfortunately they are not exception safe, I am adamant about exceptions everywhere and no return codes.    We will leave as is for now, but clearly we will have to revisit this.  I am  not clear on how to run the p2p tests you added?  They are not part of validate, are they?

Reply:
Yes, I agree that they should be exception-safe. I don't have a way of running each test manually, though that is a functionality I eventually want to add . p2ptest is a part of validate. I moved the code over from the old xbutil, but since your code is more efficient, we should replace my code with yours. 

If we are going to revisit this later, do you want to add a comment indicating that so that we don't forget later?

Reply:
I don't believe p2p test is part of validate, per the spec I am working off it is part of "advanced --p2p validate" ?

Reply:
I created XRT-694 to track code consolidation

Reply:
Thanks for creating a task. What spec are you referring to? In legacy xbutil we have p2p as a standalone test as well as as a part of validate